Dockerfile.ci

```diff
@@ -0,0 +1,26 @@
+FROM golang:1.12.5-stretch
+
+WORKDIR /go-control-plane
+
+# Fetch and preserve module dependencies
+ENV GOPROXY=https://gocenter.io
+COPY go.mod ./
+RUN go mod download
+
+# Fetch protoc modules
+RUN go get github.com/gogo/protobuf/protoc-gen-gogofast@v1.2.1
+RUN go get github.com/envoyproxy/protoc-gen-validate@v0.0.14
+
+# Install protoc 3.6.1
+RUN apt-get update && apt-get install unzip
+RUN wget -O /tmp/protoc.zip https://github.com/protocolbuffers/protobuf/releases/download/v3.6.1/protoc-3.6.1-linux-x86_64.zip  \
+  && unzip /tmp/protoc.zip -d /usr/local/ \
+  && rm /tmp/protoc.zip
+
+# circle does not like files in the checkout directory
+RUN rm go.mod
+RUN rm go.sum
+
+# add envoy
+# Due to https://github.com/envoyproxy/envoy/issues/6237, we have to pin envoy temporarily
+COPY --from=envoyproxy/envoy:0634e7bcbfea7317eaf805a657d048ffa27f519a /usr/local/bin/envoy /usr/local/bin/envoy
```

Comment:
wow, i didn't know you could copy from an arbitrary hash; TIL

Reply:
yeah, this is a side effect of staged builds. pretty neat.